File: java/generic/DoubleRing.java
package generic;
import java.util.Formatter;

public class DoubleRing implements IRing<DoubleRing>,
		ITrigonometric<DoubleRing>, IInvertible<DoubleRing>, IMath<DoubleRing>,
		IOrdered<DoubleRing> {
	double d;

	public boolean printShort = true;

	public static int fCount;

	DoubleRing(double d) {
		this.d = d;
	}

	public DoubleRing copy() {
		return new DoubleRing(d);
	}

	public DoubleRing[] newArray(int size) {
		return new DoubleRing[size];
	}

	public DoubleRing a(DoubleRing o) {
		fCount++;
		if (o == null)
			return new DoubleRing(d);
		else
			return new DoubleRing(d + o.d);
	}

	public void ae(DoubleRing o) {
		fCount++;
		if (o != null)
			d += o.d;
	}

	public DoubleRing s(DoubleRing o) {
		fCount++;
		if (o != null)
			return new DoubleRing(d - o.d);
		else
			return new DoubleRing(d);
	}

	public void se(DoubleRing o) {
		fCount++;
		if (o != null)
			d -= o.d;
	}

	public DoubleRing m(DoubleRing o) {
		fCount++;
		if (o != null)
			return new DoubleRing(d * o.d);
		else
		return new DoubleRing(0);
	}

	public void me(DoubleRing o) {
		fCount++;
		if (o != null)
			d *= o.d;
		else
			d = 0;
	}


	public DoubleRing coerce(int i) {
		return new DoubleRing(i);
	}

	public DoubleRing coerce(double d) {
		return new DoubleRing(d);
	}

	public double coerce() {
		return d;
	}

	public void sqrt() {
		d = Math.sqrt(d);
	}

	public void sin() {
		d = Math.sin(d);
	}

	public void cos() {
		d = Math.cos(d);
	}

 	public String toString() {
		if (printShort) {
			try (Formatter fmt = new Formatter()) {
				fmt.format("%6.2f", d);
				return fmt.toString();
			}
		} else {
			return Double.toString(d);
		}
	} 

	public DoubleRing newInstance() {
		return new DoubleRing(0);
	}

	public int intValue() {
		return (int) d;
	}

	public boolean isZero() {
		return d == 0.0;
	}

	public boolean isOne() {
		return d == 1.0;
	}

	public DoubleRing zero() {
		return new DoubleRing(0);
	}

	public DoubleRing one() {
		return new DoubleRing(1);
	}

	public DoubleRing invert() {
		fCount++;
		return new DoubleRing(1 / d);
	}

	public DoubleRing abs() {
		if (d < 0)
			return new DoubleRing(-d);
		return new DoubleRing(d);
	}

	public boolean lt(DoubleRing o) {
		if (d < o.d)
			return true;
		return false;
	}

	public boolean le(DoubleRing o) {
		if (d <= o.d)
			return true;
		return false;
	}

	public boolean gt(DoubleRing o) {
		if (d > o.d)
			return true;
		return false;
	}

	public boolean ge(DoubleRing o) {
		if (d >= o.d)
			return true;
		return false;
	}
}

File: rust/src/generic/double_ring.rs
use std::fmt;
use crate::generic::i_ring::IRing;
use crate::generic::i_invertible::IInvertible;
use crate::generic::i_math::IMath;
use crate::generic::i_ordered::IOrdered;
use crate::generic::i_trigonometric::ITrigonometric;
pub struct DoubleRing {
    pub d: f64,
    pub print_short: bool,
}

impl DoubleRing {
    //pub static mut FCOUNT: i32 = 0;

    pub fn new(d: f64) -> Self {
        DoubleRing { d, print_short: true }
    }

    pub fn copy(&self) -> DoubleRing {
        DoubleRing::new(self.d)
    }

    /* pub fn new_array(size: usize) -> Vec<DoubleRing> {
        vec![DoubleRing::new(0.0); size]
    } */
}


impl IRing for DoubleRing {
    fn a(&self, o: &DoubleRing) -> DoubleRing {
        //unsafe { DoubleRing::FCOUNT += 1; }
        DoubleRing::new(self.d + o.d)
    }

    fn ae(&mut self, o: &DoubleRing) {
        //unsafe { DoubleRing::FCOUNT += 1; }
        self.d += o.d;

    }

    fn s(&self, o: &DoubleRing) -> DoubleRing {
        //unsafe { DoubleRing::FCOUNT += 1; }
        DoubleRing::new(self.d - o.d)
    }

    fn se(&mut self, o: &DoubleRing) {
        //unsafe { DoubleRing::FCOUNT += 1; }
        self.d -= o.d;
    }

    fn m(&self, o: &DoubleRing) -> DoubleRing {
        //unsafe { DoubleRing::FCOUNT += 1; }
        DoubleRing::new(self.d * o.d)
    }

    fn me(&mut self, o: &DoubleRing) {
        //unsafe { DoubleRing::FCOUNT += 1; }
        self.d *= o.d;
    }

    fn coerce_from_i32(&self, i: i32) -> DoubleRing {
        DoubleRing::new(i as f64)
    }

    fn coerce_from_f64(&self, d: f64) -> DoubleRing {
        DoubleRing::new(d)
    }

    fn coerce(&self) -> f64 {
        self.d
    }

    fn is_zero(&self) -> bool {
        self.d == 0.0
    }

    fn is_one(&self) -> bool {
        self.d == 1.0
    }

    fn zero() -> DoubleRing {
        DoubleRing::new(0.0)
    }

    fn one() -> DoubleRing {
        DoubleRing::new(1.0)
    }

    fn copy(&self) -> DoubleRing {
        DoubleRing::new(self.d)
    }
}

impl IInvertible for DoubleRing {
    fn invert(&self) -> DoubleRing {
        //unsafe { DoubleRing::FCOUNT += 1; }
        DoubleRing::new(1.0 / self.d)
    }
}

impl IMath for DoubleRing {
    fn abs(&self) -> DoubleRing {
        DoubleRing::new(self.d.abs())
    }
    fn sqrt(&mut self) {
        self.d = self.d.sqrt();
    }
}

impl IOrdered for DoubleRing {
    fn lt(&self, o: &DoubleRing) -> bool {
        self.d < o.d
    }

    fn le(&self, o: &DoubleRing) -> bool {
        self.d <= o.d
    }

    fn gt(&self, o: &DoubleRing) -> bool {
        self.d > o.d
    }

    fn ge(&self, o: &DoubleRing) -> bool {
        self.d >= o.d
    }
}

impl ITrigonometric for DoubleRing {
    fn sin(&mut self) {
        self.d = self.d.sin();
    }

    fn cos(&mut self) {
        self.d = self.d.cos();
    }
}

impl fmt::Display for DoubleRing {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        if self.print_short {
            write!(f, "{:6.2}", self.d)
        } else {
            write!(f, "{}", self.d)
        }
    }
}